synapse/storage/databases/main/events_worker.py

```diff
@@ -736,35 +736,80 @@ async def get_stripped_room_state_from_event_context(
             for e in state_to_include.values()
         ]
 
+    def _start_fetch_thread(self) -> None:
+        """Starts an event fetcher."""
+        with self._event_fetch_lock:
+            if (
+                self._event_fetch_list
+                and self._event_fetch_ongoing < EVENT_QUEUE_THREADS
+            ):
+                self._event_fetch_ongoing += 1
+                event_fetch_ongoing_gauge.set(self._event_fetch_ongoing)
+                should_start = True
+            else:
+                should_start = False
+
+        async def _do_fetch() -> None:
+            """Services requests for events from the `_event_fetch_list` queue."""
+            exc = None
+            try:
+                await self.db_pool.runWithConnection(self._do_fetch)
+            except Exception as e:
+                exc = e
+                raise
+            finally:
+                should_restart = False
+                failed_event_list = []
+                with self._event_fetch_lock:
+                    self._event_fetch_ongoing -= 1
+                    event_fetch_ongoing_gauge.set(self._event_fetch_ongoing)
+
+                    if self._event_fetch_ongoing == 0 and self._event_fetch_list:
+                        # We are the last remaining fetcher and we are about to
+                        # go away. Deal with any outstanding fetches.
+                        if exc is None:
+                            should_restart = True
+                        else:
+                            failed_event_list = self._event_fetch_list
+                            self._event_fetch_list = []
+
+                if should_restart:
+                    self._start_fetch_thread()
+
+                if exc is not None:
+                    for _, deferred in failed_event_list:
+                        if not deferred.called:
+                            with PreserveLoggingContext():
```

Comment:
I'd be tempted to hoist this outside the loop. (I'm not sure it's needed at all, but it's certainly not needed for each iteration.)